src/Native/Runtime/arm/PInvoke.S

```diff
@@ -3,39 +3,77 @@
 // See the LICENSE file in the project root for more information.
 
 #include <unixasmmacros.inc>
+#include <AsmOffsets.inc>
 
 .syntax unified
 .thumb
 
-#ifdef _DEBUG
-.rodata
-AssertMsg:                              .asciz "__FILE__:%s: %s is not implemented\n"
-FileName:                               .asciz "PInvoke.S"
-RhpPInvokeName:                         .asciz "RhpPInvoke"
-RhpPInvokeReturnName:                   .asciz "RhpPInvokeReturn"
+.macro INLINE_GET_TLS_VAR Var, Reg
+	        ldr	         r0, =\Var@TLSLD
+	        bl	         __tls_get_addr@PLT
+	        ldr	         \Reg, =\Var@DTPOFF
+	        add	         r0, \Reg
+.endmacro
 
-.text
-.macro GEN_ASSERT_FUNC func
-        GEN_ASSERT AssertMsg, FileName, \func
-.endm
-#endif
-
-// TODO: Implement Arm support
+.macro INLINE_GETTHREAD Reg
+	        // Inlined version of call C_FUNC(RhpGetThread)
+	        INLINE_GET_TLS_VAR tls_CurrentThread, \Reg
+.endmacro
 
 //
 // RhpPInvoke
 //
+// IN:  R0: address of pinvoke frame
+//
+// This helper assumes that its callsite is as good to start the stackwalk as the actual PInvoke callsite.
+// The codegenerator must treat the callsite of this helper as GC triggering and generate the GC info for it.
+// Also, the codegenerator must ensure that there are no live GC references in callee saved registers.
+//
 NESTED_ENTRY RhpPInvoke, _TEXT, NoHandler
-#ifdef _DEBUG
-        GEN_ASSERT_FUNC RhpPInvokeName
-#endif
+	        PROLOG_PUSH "{r5-r7,lr}"
+
+	        mov	        r5, r0
+
+	        // R0 = GetThread(), R6 - Trash Register
+	        INLINE_GETTHREAD r6
+
+	        str         r0, [r5, #OFFSETOF__PInvokeTransitionFrame__m_pThread]
+	        str         r7, [r5, #OFFSETOF__PInvokeTransitionFrame__m_FramePointer]
+	        str         r14, [r5, #OFFSETOF__PInvokeTransitionFrame__m_RIP]
```

Comment:
r14 will be trashed by the `__tls_get_addr@PLT ` call above. You need to save it before.